File: solutions/ruby/robot-simulator/robot_simulator.rb
class Robot
  private

  attr_writer :bearing, :coordinates
  attr_accessor :pos_x, :pos_y

  private_constant

  BEARINGS = %i[east north west south].freeze
  OFFSETS = {
    east: [1, 0],
    north: [0, 1],
    west: [-1, 0],
    south: [0, -1]
  }.freeze

  def initialize
    at(0, 0)
    orient(:east)
  end

  public

  attr_reader :bearing, :coordinates

  def at(pos_x, pos_y)
    self.coordinates = [pos_x, pos_y]
  end

  def orient(orientation)
    raise ArgumentError, 'Invalid bearing' unless BEARINGS.include?(orientation)

    self.bearing = orientation
  end

  def turn_right
    self.bearing = BEARINGS[(BEARINGS.index(bearing) + 3) % 4]
  end

  def turn_left
    self.bearing = BEARINGS[(BEARINGS.index(bearing) + 1) % 4]
  end

  def advance
    self.coordinates = coordinates.zip(OFFSETS[bearing]).map(&:sum)
  end
end

class Simulator
  private_constant

  COMMANDS = { 'R' => :turn_right, 'L' => :turn_left, 'A' => :advance }.freeze

  def place(robot, x: nil, y: nil, direction: nil)
    robot.at(x, y) unless x.nil? || y.nil?
    robot.orient(direction) unless direction.nil?
  end

  def instructions(commands)
    commands.chars.map { |char| COMMANDS[char] }
  end

  def evaluate(robot, commands)
    instructions(commands).each { |command| robot.send(command) }
  end
end

File: solutions/python/robot-simulator/robot_simulator.py
"""Robot Simulator."""

# Directions, complex polar mappings.
EAST, NORTH, WEST, SOUTH = (1j ** n for n in range(4))
ROTATE = {"R": -1j, "L": 1j}


class Robot:
    """A robot."""

    def __init__(self, direction: complex, x: int, y: int):
        """Initialize, mapping inputs to complex numbers."""
        self.direction = direction
        self._coord = complex(x, y)

    @property
    def coordinates(self) -> tuple[int, int]:
        """Map complex coordinates to Cartesian."""
        return (int(self._coord.real), int(self._coord.imag))

    def move(self, instructions: str) -> None:
        """Move the robot Right|Left|Advance."""
        for instruction in instructions:
            if instruction == "A":
                # Advance.
                self._coord += self.direction
            elif instruction in ROTATE:
                self.direction *= ROTATE[instruction]
            else:
                raise ValueError(f"invalid instruction {instruction}")
